### PR TITLE
SONARJAVA-5410 Modify rule S5977: Improve Rationale for Avoiding Randomness in Tests

### PR DESCRIPTION
**Pull request by tomasz-tylenda-sonarsource**
[SONARJAVA-5410](https://sonarsource.atlassian.net/browse/SONARJAVA-5410)

[SONARJAVA-5410]: https://sonarsource.atlassian.net/browse/SONARJAVA-5410?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rules/S5977/java/rule.adoc [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
+* Flaky tests, which pass or fail intermittently without any code changes, are a significant problem for CI pipelines (continuous integration).
+They erode confidence in the CI system, lead to unnecessary investigations and reruns, and ultimately slow down the development and release process.
+A stable CI pipeline relies on deterministic test outcomes.
 
+This rule raises an issue when ``++new Random()++`` or ``++UUID.randomUUID()++`` are called in test code.
 
-There is one valid use case for random data in tests: when testing every value would make tests impractically slow. In this case the best you can do is use random to test every value on the long run. You should however make sure that random values are logged so that you can reproduce failures. Some libraries exist to make all this easier. You can for example use property-based testing libraries such as https://github.com/jlink/jqwik[jqwik].
+=== Exceptions
 
+This rule excludes the use of library functions that read the system clock (`++new Date()++`, `++Instant.now()++`, and so on).
```
Comment (dorian-burihabwa-sonarsource) [inline rules/S5977/java/rule.adoc:21]: There wasn't any exception section before. Is this something we see from the source code?

Comment (dorian-burihabwa-sonarsource) [inline rules/S5977/java/rule.adoc:21]:
```suggestion
This rule excludes the use of library functions that read the system clock (e.g., `++new Date()++`, `++Instant.now()++`, ...).
```

File: rules/S5977/java/rule.adoc [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
 
-This rule raises an issue when ``++new Random()++`` or ``++UUID.randomUUID()++`` are called in test code.
+== How to Fix?
+
+* When a test uses random numbers to generate inputs, an easy fix is to replace those random inputs with pseudo-random values generated from a known seed.
+By initializing a pseudo-random number generator with a fixed seed, tests can generate sequences of seemingly random data that are entirely reproducible across different test runs.
```
Comment (dorian-burihabwa-sonarsource) [inline rules/S5977/java/rule.adoc:26]:
```suggestion
By initializing a pseudo-random number generator with a fixed seed, tests can generate sequences of seemingly random data that are reproducible across different test runs.
```

File: rules/S5977/java/rule.adoc [hunk 3]
```diff
@@ -1,17 +1,32 @@
...
-Tests should always:
+Randomness in test code, whether introduced intentionally to cover multiple scenarios or unintentionally through non-deterministic library functions, undermines the principles of reliable and effective testing.
+In almost all cases, it leads to problems and should be avoided.
 
-* Make sure that production code behaves as expected, including edge cases.
-* Be easy to debug, i.e. understandable and reproducible.
+Fundamentally, we aim to make tests deterministic and reproducible, primarily for the following reasons:
 
-Using random values in tests will not necessarily check edge cases, and it will make test logs a lot harder to read. It is better to use easily readable hardcoded values. If this makes your code bigger you can use helper functions.
+* When a test fails, the ability to reproduce the exact conditions that led to the failure is crucial for effective debugging.
```
Comment (dorian-burihabwa-sonarsource) [inline rules/S5977/java/rule.adoc:8]:
```suggestion
* When a test fails, the ability to reproduce the conditions that led to the failure is crucial for effective debugging.
```

File: rules/S5977/java/rule.adoc (hunk 1 above)
Comment (tomasz-tylenda-sonarsource) [reply rules/S5977/java/rule.adoc]: I removed it.

It was meant to discuss another source of randomness, which is not in scope of this rule. I decided it will be better to remove it to avoid confusion.
